### PR TITLE
[#7215] Platform: Ability to select and delete multiple backups.

### PR DESCRIPTION
**Pull request by nishantSharma459**
Providing the users such an environment where they'll have access to delete the multiple backups at once.

MutipleBackupDelete

sbt test : `[info] Passed: Total 1266, Failed 0, Errors 0, Passed 1262, Skipped 4
[success] Total time: 919 s (15:19), completed 1 Apr, 2021 2:27:21 PM`

### CODE REVIEW

File: managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java [hunk 1]
```diff
@@ -163,37 +165,47 @@ public Result restore(UUID customerUUID, UUID universeUUID) {
...
-        CustomerTask.TaskType.Delete,
-        "Backup");
     ObjectNode resultNode = Json.newObject();
-    resultNode.put("taskUUID", taskUUID.toString());
-    Audit.createAuditEntry(ctx(), request(), taskUUID);
+    for (UUID uuid : validBackups) {
+      Backup backup = Backup.get(customerUUID, uuid);
+      
+      if (backup.state != Backup.BackupState.Completed) {
+        LOG.info("Can not delete {} backup as it is already in progress", uuid);
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java:192]: Not `already` but `still`.

File: managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java [hunk 2]
```diff
@@ -163,37 +165,47 @@ public Result restore(UUID customerUUID, UUID universeUUID) {
...
-    CustomerTask.create(customer,
-        backupUUID,
-        taskUUID,
-        CustomerTask.TargetType.Backup,
-        CustomerTask.TaskType.Delete,
-        "Backup");
     ObjectNode resultNode = Json.newObject();
-    resultNode.put("taskUUID", taskUUID.toString());
-    Audit.createAuditEntry(ctx(), request(), taskUUID);
+    for (UUID uuid : validBackups) {
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java:188]: You have split one loop into two. And now for each existing backup you are searching it twice (using Backup.get()). You can fix this using one loop or changing
`List<UUID> validBackups = new ArrayList<>();
`
to
`List<Backup> validBackups = new ArrayList<>();`

File: managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java [hunk 3]
```diff
@@ -163,37 +165,47 @@ public Result restore(UUID customerUUID, UUID universeUUID) {
     return ApiResponse.success(resultNode);
   }
 
-  public Result delete(UUID customerUUID, UUID backupUUID) {
+  public Result delete(UUID customerUUID) {
+    List<UUID> validBackups = new ArrayList<UUID>();
```
Comment (SergeyPotachev) [inline managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java:169]: Don't use type specifiers in the right part:
```
List<UUID> validBackups = new ArrayList<>();
List<String> taskUUIDList = new ArrayList<>();
```

Comment (SergeyPotachev) [reply managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java:169]: Also move this block after the `(customer == null)` check.

File: managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java (hunk 1 above)
Comment (SergeyPotachev) [reply managed/src/main/java/com/yugabyte/yw/controllers/BackupsController.java:192]: Also, as I've already said about two loops instead on the initial one, this check should be moved to the first loop.

File: managed/src/test/java/com/yugabyte/yw/controllers/BackupsControllerTest.java [hunk 4]
```diff
@@ -84,6 +88,17 @@ private Result restoreBackup(UUID universeUUID, JsonNode bodyJson, Users user) {
     return FakeApiHelper.doRequestWithAuthTokenAndBody(method, url, authToken, bodyJson);
   }
 
+  private Result deleteBackup(ObjectNode bodyJson, Users user) {
+    String authToken = defaultUser.createAuthToken();
+    if (user != null) {
+      authToken = user.createAuthToken();
+    }
+    String method = "DELETE";
```
Comment (SergeyPotachev) [inline managed/src/test/java/com/yugabyte/yw/controllers/BackupsControllerTest.java:96]: `String authToken = user == null ? defaultUser.createAuthToken() : user.createAuthToken();`

File: managed/src/test/java/com/yugabyte/yw/controllers/BackupsControllerTest.java [hunk 5]
```diff
@@ -241,4 +256,28 @@ public void testRestoreBackupRequestTooLarge() {
...
   }
+
+  @Test
+  public void testDeleteBackup() {
+    CustomerConfig customerConfig = ModelFactory.createS3StorageConfig(defaultCustomer);
+    BackupTableParams bp = new BackupTableParams();
+    bp.storageConfigUUID = customerConfig.configUUID;
+    Backup backup = Backup.create(defaultCustomer.uuid, bp);
+    backup.transitionState(BackupState.Completed);
+    List<String> backupUUIDList = new ArrayList<String>();
```
Comment (SergeyPotachev) [inline managed/src/test/java/com/yugabyte/yw/controllers/BackupsControllerTest.java:267]: ` List<String> backupUUIDList = new ArrayList<>();`

File: managed/src/test/java/com/yugabyte/yw/controllers/BackupsControllerTest.java [hunk 6]
```diff
@@ -241,4 +256,28 @@ public void testRestoreBackupRequestTooLarge() {
...
+    backupUUIDList.add(backup.backupUUID.toString());
+    UUID fakeTaskUUID = UUID.randomUUID();
```
Comment (SergeyPotachev) [inline managed/src/test/java/com/yugabyte/yw/controllers/BackupsControllerTest.java:269]: You don't need this variable.

**Review by SergeyPotachev — CHANGES_REQUESTED** [review]
Please see comments related to BE.

File: managed/ui/src/components/tables/DeleteBackup/DeleteBackup.js [hunk 7]
```diff
@@ -12,20 +12,29 @@ export default class DeleteBackup extends Component {
...
+      },
       deleteBackup,
       onHide,
       onSubmit,
       onError
-    } = this.props;
+    } = this.props;   
+    let payload = [];
+
+    type === "bulkDelete" ? payload = data : payload.push(backupUUID);
```
Comment (yb-andrew) [inline managed/ui/src/components/tables/DeleteBackup/DeleteBackup.js:27]: Why not do `const payload = type === 'bulkDelete' ? data : [backupUUID]`?

File: managed/ui/src/components/tables/ListBackups/ListBackups.js [hunk 8]
```diff
@@ -283,6 +285,42 @@ export default class ListBackups extends Component {
...
+   *
+   * @param {boolean} isSelected Selected rows.
+   * @param {Array} rows Number of rows.
+   * @returns Boolean
+   */
+  onSelectAll = (isSelected, rows) => {
+    let selected = [...this.state.selected];
+    if (isSelected) {
+      rows.forEach((row) => selected.push(row.backupUUID));
+      this.setState({ selected: selected });
```
Comment (yb-andrew) [inline managed/ui/src/components/tables/ListBackups/ListBackups.js:316]: Just move the variable declaration into the if-block: `const selected = [...this.state.selected, ...rows.map(row => row.backupUUID)]`

File: managed/ui/src/components/tables/ListBackups/ListBackups.js [hunk 9]
```diff
@@ -283,6 +285,42 @@ export default class ListBackups extends Component {
...
+   * @param {string} {backupUUID} Backup UUID.
+   * @param {boolean} isSelected Selected row.
+   * @returns Boolean
+   */
+  onRowSelect = ({ backupUUID }, isSelected) => {
+    isSelected ?
+      this.setState({ selected: [...this.state.selected, backupUUID].sort() }) :
+      this.setState({ selected: this.state.selected.filter((id) => id !== backupUUID) });
+
+    return false;
```
Comment (yb-andrew) [inline managed/ui/src/components/tables/ListBackups/ListBackups.js:301]: Return value should most likely be true even if you are using state to select the row.

File: managed/ui/src/components/tables/ListBackups/ListBackups.js [hunk 10]
```diff
@@ -405,13 +454,16 @@ export default class ListBackups extends Component {
         )}
         {showAlert && (
           <Alert bsStyle={taskUUID ? 'success' : 'danger'} onDismiss={this.handleDismissAlert}>
-            {taskUUID ? (
+            {taskUUID.length < 2 ? (
```
Comment (yb-andrew) [inline managed/ui/src/components/tables/ListBackups/ListBackups.js:457]: Check that taskUUID is array

**Review by yb-andrew — CHANGES_REQUESTED** [review]

File: managed/ui/src/components/tables/DeleteBackup/DeleteBackup.js (hunk 7 above)
Comment (nishantSharma459) [reply managed/ui/src/components/tables/DeleteBackup/DeleteBackup.js]: Didn't think of it this way but implemented the suggested one.

File: managed/ui/src/components/tables/ListBackups/ListBackups.js (hunk 8 above)
Comment (nishantSharma459) [reply managed/ui/src/components/tables/ListBackups/ListBackups.js]: Done

File: managed/ui/src/components/tables/ListBackups/ListBackups.js (hunk 9 above)
Comment (nishantSharma459) [reply managed/ui/src/components/tables/ListBackups/ListBackups.js]: Done

File: managed/ui/src/components/tables/ListBackups/ListBackups.js (hunk 10 above)
Comment (nishantSharma459) [reply managed/ui/src/components/tables/ListBackups/ListBackups.js]: Done

**Review by SergeyPotachev — APPROVED** [review]
The backend part looks good to me.

**Review by yb-andrew — APPROVED** [review]
LG! Thanks.

**Pull request merged**